src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -0,0 +1,292 @@
+        {
+            _session = session;
+            _scope = scope;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void AddAdditionalAssets(CustomAsset asset)
+        {
+            _scope.AddAdditionalAsset(asset, _cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            _scope.Dispose();
+            _session.Dispose();
+        }
+
+        public Task InvokeAsync(string targetName, params object[] arguments) =>
+            _session.InvokeAsync(targetName, arguments);
+        public Task<T> InvokeAsync<T>(string targetName, params object[] arguments) =>
+            _session.InvokeAsync<T>(targetName, arguments);
+        public Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync);
+        public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync);
+    }
+
+    /// <summary>
+    /// Helper type for common session case
+    /// 
+    /// This will let one to hold onto <see cref="RemoteHostClient.Session"/> for a while.
+    /// this helper will let you not care about remote host being gone while you hold onto the session if that ever happen
+    /// 
+    /// and also make sure state is correct even if multiple threads call TryInvokeAsync at the same time. but this 
+    /// is not optimized to handle highly concurrent usage. if highly concurrent usage is required, either using
+    /// <see cref="RemoteHostClient.Session"/> direclty or using <see cref="SolutionAndSessionHolder"/> would be better choice
+    /// </summary>
+    internal sealed class KeepAliveSessionHolder
+    {
+        private readonly SemaphoreSlim _gate;
+        private readonly IRemoteHostClientService _remoteHostClientService;
+        private readonly CancellationToken _cancellationToken;
+
+        private readonly string _serviceName;
+        private readonly object _callbackTarget;
+
+        private RemoteHostClient _client;
+        private RemoteHostClient.Session _session;
+
+        public KeepAliveSessionHolder(RemoteHostClient client, RemoteHostClient.Session session, string serviceName, object callbackTarget, CancellationToken cancellationToken)
+        {
+            Initialize_NoLock(client, session);
+
+            _gate = new SemaphoreSlim(initialCount: 1);
+            _remoteHostClientService = client.Workspace.Services.GetService<IRemoteHostClientService>();
+            _cancellationToken = cancellationToken;
+
+            _serviceName = serviceName;
+            _callbackTarget = callbackTarget;
+        }
+
+        public void Shutdown()
```

Comment:
> but it doesn't. since their behavior is different

Their behavior isn't different.  I outlined above how all the feature-layer clients use these types.  They get the session.  The call operations on it.  They dispose the session once they're done with it.

> I am not sure why you are trying to make those sit on same inheritance chain.

To me, it's because these all seem to be the same type.  And that's evident in that you have to expose the same methods on all your new types (which just end up doing forwarding).  Basically, you've introduced new types, that all have the same shape, with the justification that the client needs to have that different shape so they are aware of the distinction.  But my point is that as the client (which i am through all the features that i've written), i don't really benefit or care about this implementation detail leaking through.  To me, at the end of the day it's just a 'session'.  I get it configured based on parameters i pass you.  I call into it.  And i clean it up when i'm done with it.